### PR TITLE
Add a default set of semgrepignore patterns

### PR DESCRIPTION
**Pull request by mmcqd**
Now the file `semgrep/templates/.semgrepignore` will be used as a set of patterns to ignore by default if no `.semgrepignore` file is found in the directory where Semgrep is being run. This PR also moves the `.semgrepignore` filtering code to the `filtered_files` function, so that it is cached on a per language basis.

test plan: make test (see added test)

PR checklist:
- [x] Documentation is up-to-date
- [x] Changelog is up-to-date
- [x] Change has no security implications (otherwise, ping security team)

### CODE REVIEW

**Comment by mmcqd** [comment]
Seems like this causes some significant slowdowns.

**Comment by mmcqd** [comment]
Failing because the `templates` dir doesn't get included in the docker image. Could be solved by just storing the default `.semgrepignore` as a string?

**Comment by emjin** [comment]
hm these slowdowns are worse than I realized. Most of these repos are pretty small. Can you time some larger repos?

**Comment by emjin** [comment]
Whoo the benchmark tests work, always good to see. If you can't figure out in like 5 minutes how to fix the benchmark tests, ping me to document this better.

**Review by emjin — APPROVED** [review]

File: semgrep/semgrep/semgrep_main.py [hunk 1]
```diff
@@ -63,21 +63,21 @@ def notify_user_of_work(
...
-            )
+    if not semgrepignore_path.is_file():
+        logger.verbose("using default .semgrepignore")
+        semgrepignore_path = TEMPLATES_DIR / IGNORE_FILE_NAME
     else:
-        logger.verbose("no .semgrepignore found")
-        file_ignore = None
+        logger.verbose("using path ignore rules from user provided .semgrepignore")
+
+    with open(semgrepignore_path) as f:
```
Comment (brendongo) [inline semgrep/semgrep/semgrep_main.py]:
```suggestion
    with semgrepignore_path.open() as f:
```
Path objects expose an open helper :D

File: semgrep/semgrep/semgrep_main.py [hunk 2]
```diff
@@ -63,21 +63,21 @@ def notify_user_of_work(
...
     semgrepignore_path = Path(workdir / IGNORE_FILE_NAME)
-    if semgrepignore_path.is_file():
-        logger.verbose("using path ignore rules from .semgrepignore")
-        with open(semgrepignore_path) as f:
-            file_ignore: Optional[FileIgnore] = FileIgnore(
-                base_path=workdir,
-                patterns=Parser(workdir).parse(f),
-            )
+    if not semgrepignore_path.is_file():
+        logger.verbose("using default .semgrepignore")
```
Comment (brendongo) [inline semgrep/semgrep/semgrep_main.py]:
```suggestion
        logger.verbose("No .semgrepignore found. Using default ignores rules.")
```
Thoughts?

**Review by brendongo — APPROVED** [review]
Some nits.

File: semgrep/semgrep/semgrep_main.py (hunk 2 above)
Comment (ievans) [reply semgrep/semgrep/semgrep_main.py]: can we provide a pointer to it? maybe a link to docs?

**Comment by emjin** [comment]
Collllleeeeeeeeeeen

Edit: nvm I can't read

File: semgrep/semgrep/semgrep_main.py (hunk 2 above)
Comment (mmcqd) [reply semgrep/semgrep/semgrep_main.py]: Will update this and the docs later with info about the default.

Comment (mmcqd) [reply semgrep/semgrep/semgrep_main.py]: For now just made Brendon's change

**Pull request merged**